### ISSUE TITLE
Block sync problem in short block time chain.

### ISSUE THREAD

**Issue by beer-1** [cometbft/cometbft#3340]
### Setup

**CometBFT version** (use `cometbft version` or `git rev-parse --verify HEAD` if installed from source): v0.38.x

**Have you tried the latest version**: yes

**ABCI app** (name for built-in, URL for self-written if it's publicly available): initia, injective, sei

### What happened?

We are currently running single validator network with 0.5s block time setting and trying to keep syncing from other full node.
But periodically, the full node keeps failing to keep sync and fall behind. and wired thing is that restarting the node solves the problem.

And also got some reports like they also having this problem in Sei and Injective chain and solve the issue by restarting the node.

### What did you expect to happen?

no problem in block sync

### How to reproduce it

Prease

**Comment by ValarDragon** [comment]
Yeah, I think our problem is that in consensus mode:
- Nodes don't save data sent from future heights/rounds
- Peers as a consequence try to not gossip things too far in the future
- Once a peer has sent a vote/block part they will never send it again.

**Comment by cason** [comment]
I think one of the problems here is the following. To decide a block we need 2/3+ Precommits for the block, they contain the `BlockID`, and we need the full block matching that `BlockID`.

When a node realizes that a peer is lagging behind it does the follow, in different routines:

1. Sends Precommits for the `BlockID` of the block committed at the reported peer height
2. Sends block Parts for the `Block` committed at the reported peer height

We only accept a block Part if it matches our `cs.ProposalBlockParts`, which is a Merkle tree. We setup this Merkle tree upon receiving a `Proposal`, this is the ordinary operation, **or** when we see 2/3+ Prevotes or Precommits for a block (that is not the block we are building or have built).

What matters here is the second case with Precommit messages. Only when we get the 2/3+ Precommits for a `BlockID`, we start accepting the Parts of the associated block:

https://github.com/cometbft/cometbft/blob/4d98dc1b2f611e8f2545172e66799a7bd9d7a317/internal/consensus/state.go#L1774-L1786

In summary, we are receiving block parts before we receive the 2/3+ Precommits that would allow us to accept the associated block parts. This means that the block parts that we have received, although good ones, were dropped, and we are waiting for their duplicates, eventually coming from other peers.

To attest my thesis, you could look to the INFO message of the code above and increase the log level of the message produced when we reject a block part:

https://github.com/cometbft/cometbft/blob/4d98dc1b2f611e8f2545172e66799a7bd9d7a317/internal/consensus/state.go#L2125-L2138

If my thesis is right, we first reject parts, then get the 2/3+ precommits, then start accepting block parts, then build the full block and eventually commit.

**Comment by cason** [comment]
We should probably rephrase the issue.

The problem is on slow/catching-up peers in consensus. The solution is to restart the node, so that it performs block sync.
